modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -854,4 +891,66 @@ static boolean isRecoverable(Throwable throwable) {
             return false;
         }
     }
+
+    /**
+     * Handles messages of {@link TxMessageGroup}.
+     *
+     * @see NetworkMessageHandler#onReceived(NetworkMessage, String, Long)
+     */
+    private void onReceiveTxNetworkMessage(NetworkMessage message, String senderConsistentId, @Nullable Long correlationId) {
+        if (!(message instanceof RwTransactionsFinishedRequest)) {
+            return;
+        }
+
+        assert correlationId != null : senderConsistentId;
+
+        int targetCatalogVersion = ((RwTransactionsFinishedRequest) message).targetSchemaVersion();
+
+        boolean finished = isRwTransactionsFinished(targetCatalogVersion);
+
+        clusterService.messagingService().respond(senderConsistentId, toRwTransactionsFinishedResponse(finished), correlationId);
+    }
+
+    /**
+     * Returns the transaction catalog version.
+     *
+     * @param beginTs Transaction begin timestamp.
+     */
+    private int txCatalogVersion(HybridTimestamp beginTs) {
+        return catalogService.activeCatalogVersion(beginTs.longValue());
+    }
+
+    /**
+     * Returns the transaction catalog version.
+     *
+     * @param txId Transaction ID.
+     */
+    private int txCatalogVersion(UUID txId) {
+        return txCatalogVersion(TransactionIds.beginTimestamp(txId));
+    }
+
+    /**
+     * Returns {@code true} iff the requested catalog version is registered locally and all RW transactions before it have completed,
+     * otherwise {@code false}.
+     *
+     * @param catalogVersion Catalog version to check.
+     */
+    boolean isRwTransactionsFinished(int catalogVersion) {
+        if (catalogVersion > catalogService.latestCatalogVersion()) {
+            // Requested catalog version has not yet registered locally.
+            return false;
+        }
+
+        return !rwTxCounterByCatalogVersion.isExistsTxBefore(catalogVersion);
+    }
+
+    private void decrementRwTxCount(TxStateMeta txStateMeta, UUID txId) {
```

Comment:
This new logic about maintaining counters and responding to messages increased `TxManagerImpl` quite a bit. It seems that it might make sense to create another class just to encapsulate this logic and put all the new methods and fields in it, to make sure this logic does not get in the way of perceiving the main transaction-related logic. WDYT?

Reply:
Whu not